### PR TITLE
Attempts to fix exception handlers added by emscripten.

### LINKED ISSUE #34: Improve greedy exception catcher of Emscripten
Follow up from https://github.com/chriseth/browser-solidity/issues/167.

The Emscripten preamble inserts an exception catcher for _uncaughtException_ in Node.js.  This rethrows all exceptions, bar the ExitCode one.

The rethrow causes stack traces to be confusing long.

### LINKED ISSUE #84: `solc` breaks mocha test runner.
Repro case here: https://github.com/MicahZoltu/mocha-test/

Run either with `docker build -t mocha-test . && docker run --rm mocha-test` or with `npm install && node_modules/.bin/mocha`.  Notice that the test runner crashes.

Comment out the line here https://github.com/MicahZoltu/mocha-test/blob/master/test/example.js#L2-L3 and then notice that the test runner does not crash.

Prime suspect is https://github.com/ethereum/solc-js/blob/master/solcjs#L8, though replacing the `require("solc")` line with that does not cause the problem to persist.  I have been unable to narrow down the problem past `solc`, though the test runner crash output suggests that `solc` is to blame.

### PR DESCRIPTION
**Pull request by MicahZoltu**
This (hopefully) will fix #84 and also be a little less rude to anyone who is using this library by putting things back the way they were before the script ran.

Fixes #34 and #84.

### CODE REVIEW

**Comment by MicahZoltu** [comment]
cc @axic

**Comment by chriseth** [comment]
@axic can you take a look, please?

**Comment by chriseth** [comment]
I'm restarting the tests again.

File: solcjs [hunk 1]
```diff
@@ -70,3 +70,8 @@ for (var contractName in output.contracts) {
     fs.writeFileSync(path.join(destination, contractName + '.abi'), output.contracts[contractName].interface);
   }
 }
+
+// FIXME: remove annoying exception catcher of Emscripten
+//        see https://github.com/chriseth/browser-solidity/issues/167
+process.removeAllListeners('uncaughtException');
+originalUncaughtExceptionListeners.forEach((listener) => process.addListener("uncaughtException", listener));
```
Comment (axic) [inline solcjs]: Please change to a regular function, we do not require ES6 in this project.

Comment (axic) [inline solcjs]: Also ensure to use single quotes for semistandard.

File: solcjs [hunk 2]
```diff
@@ -1,11 +1,11 @@
 #!/usr/bin/env node
 
+// hold on to any exception handlers that existed prior to this script running, we'll be adding them back at the end
+const originalUncaughtExceptionListeners = process.listeners("uncaughtException")
```
Comment (axic) [inline solcjs]: Use `var` please since this is not ES6.

File: solcjs (hunk 1 above)
Comment (MicahZoltu) [inline solcjs]: Fixed.

File: solcjs (hunk 2 above)
Comment (MicahZoltu) [inline solcjs]: Fixed.

**Comment by axic** [comment]
Thanks! Please also rebase to latest master.

File: solcjs [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
 
+// hold on to any exception handlers that existed prior to this script running, we'll be adding them back at the end
+var originalUncaughtExceptionListeners = process.listeners("uncaughtException");
+
 var fs = require('fs-extra');
 var path = require('path');
 var solc = require('./index.js');
-// FIXME: remove annoying exception catcher of Emscripten
-//        see https://github.com/chriseth/browser-solidity/issues/167
-process.removeAllListeners('uncaughtException');
```
Comment (axic) [inline solcjs]: Oh, actually the removal must stay here.

**Comment by MicahZoltu** [comment]
Applied changes suggested in PR feedback and rebased on master.

File: solcjs (hunk 3 above)
Comment (MicahZoltu) [inline solcjs]: Ah, will fix.

Comment (MicahZoltu) [inline solcjs]: Fixed.

**Pull request merged**